iree/compiler/Dialect/Flow/Transforms/MatmulToMMT4dConversion.cpp

```diff
@@ -0,0 +1,200 @@
+    };
+
+    bool isInitTensorDst = false;
+    if (dst.getDefiningOp() && isa<linalg::FillOp>(dst.getDefiningOp())) {
+      auto fillOpValue = cast<linalg::FillOp>(dst.getDefiningOp());
+      if (fillOpValue.output().getDefiningOp() &&
+          isa<linalg::InitTensorOp>(fillOpValue.output().getDefiningOp())) {
+        isInitTensorDst = true;
+      }
+    }
+    auto createTransposedInitTensor =
+        [&](ArrayRef<int64_t> targetShape) -> Value {
+      return rewriter.create<linalg::InitTensorOp>(loc, targetShape,
+                                                   dstType.getElementType());
+    };
+
+    auto lhs4D = expandTo4D(lhs, {m1, M0Size, k1, K0Size});
+    auto rhs4D = expandTo4D(rhs, {k1, K0Size, n1, N0Size});
+    auto dst4D = isInitTensorDst
+                     ? createTransposedInitTensor({m1, n1, M0Size, N0Size})
+                     : expandTo4D(dst, {m1, M0Size, n1, N0Size});
+    auto transposeOperand = [&](Value operand,
+                                ArrayRef<int64_t> indices) -> Value {
+      RankedTensorType operandTensorType =
+          operand.getType().cast<RankedTensorType>();
+      auto nloops = indices.size();
+      auto inputShape = operandTensorType.getShape();
+
+      SmallVector<AffineExpr, 4> exprs = llvm::to_vector<4>(
+          llvm::map_range(indices, [&](int64_t index) -> AffineExpr {
+            return rewriter.getAffineDimExpr(index);
+          }));
+
+      SmallVector<int64_t> targetShape = llvm::to_vector<4>(llvm::map_range(
+          indices,
+          [&](int64_t index) -> int64_t { return inputShape[index]; }));
+
+      Value outputTensor = rewriter.create<linalg::InitTensorOp>(
+          loc, targetShape, operandTensorType.getElementType());
+
+      SmallVector<StringRef> loopAttributeTypes(nloops, "parallel");
+
+      SmallVector<AffineMap> indexingMaps = {
+          inversePermutation(
+              AffineMap::get(nloops, 0, exprs, rewriter.getContext())),
+          AffineMap::getMultiDimIdentityMap(nloops, rewriter.getContext())};
+
+      auto transposedOp = rewriter.create<linalg::GenericOp>(
+          loc, outputTensor.getType(),
+          /*inputs=*/operand, /*outputs=*/outputTensor, indexingMaps,
+          loopAttributeTypes,
+          [&](OpBuilder &nestedBuilder, Location nestedLoc, ValueRange args) {
+            nestedBuilder.create<linalg::YieldOp>(nestedLoc, args[0]);
+          });
+
+      return transposedOp.getResult(0);
+    };
+
+    auto lhs4DT = transposeOperand(lhs4D, {0, 2, 1, 3});
+    auto rhs4DT = transposeOperand(rhs4D, {2, 0, 1, 3});
+    auto dst4DT =
+        isInitTensorDst ? dst4D : transposeOperand(dst4D, {0, 2, 1, 3});
+
+    auto mmt4DResult = rewriter.create<linalg::Mmt4DOp>(
+        loc, dst4DT.getType(), ValueRange{lhs4DT, rhs4DT}, ValueRange{dst4DT});
+
+    auto mmt4dResultTransposed =
+        transposeOperand(mmt4DResult.getResult(0), {0, 2, 1, 3});
+
+    auto collapseTo2D = [&](Value operand,
```

Comment:
This too seems better of being a separate function. Lambdas make it hard to read the code. Have to context switch between lambda and code outside lambda when reading the code.